src/main/UserActivityMonitor.js

```diff
@@ -0,0 +1,148 @@
+      this.forceUserIsInactive = true;
+      this.updateUserActivityStatus(false);
+    };
+    this.handleResume = () => {
+      this.forceUserIsInactive = false;
+      this.updateUserActivityStatus(true);
+    };
+    this.handleShutdown = () => {
+      this.forceUserIsInactive = true;
+      this.updateUserActivityStatus(false);
+    };
+    this.handleLockScreen = () => {
+      this.forceUserIsInactive = true;
+      this.updateUserActivityStatus(false);
+    };
+    this.handleUnlockScreen = () => {
+      this.forceUserIsInactive = false;
+      this.updateUserActivityStatus(true);
+    };
+  }
+
+  // public properties
+  get userIsActive() {
+    return this.forceUserIsInactive ? false : this.isActive;
+  }
+  get userIdleTime() {
+    return this.idleTime;
+  }
+
+  // class methods
+  // - begin monitoring system events and idle time at defined frequency
+  startMonitoring(config = {}) {
+    if (app.isReady()) {
+      // store config
+      this.config = Object.assign({}, this.config, config);
+
+      // handle system events
+      electron.powerMonitor.on('suspend', this.handleSuspend);
+      electron.powerMonitor.on('resume', this.handleResume);
+      electron.powerMonitor.on('shutdown', this.handleShutdown);
+      electron.powerMonitor.on('lock-screen', this.handleLockScreen);
+      electron.powerMonitor.on('unlock-screen', this.handleUnlockScreen);
+
+      // monitor system idle time
+      clearInterval(this.systemIdleTimeIntervalID);
+      this.systemIdleTimeIntervalID = setInterval(() => {
+        try {
+          electron.powerMonitor.querySystemIdleTime((idleTime) => {
+            this.updateIdleTime(idleTime);
+          });
+        } catch (err) {
+          console.log('Error getting system idle time:', err);
+        }
+      }, this.config.regularUpdateFrequency);
+    } else {
+      this.emit('error', new Error('UserActivityMonitor.startMonitoring can only be called after app is ready'));
+    }
+  }
+
+  // - stop monitoring system events and idle time
+  stopMonitoring() {
+    // stop monitoring system events
+    electron.powerMonitor.off('suspend', this.handleSuspend);
+    electron.powerMonitor.off('resume', this.handleResume);
+    electron.powerMonitor.off('shutdown', this.handleShutdown);
+    electron.powerMonitor.off('lock-screen', this.handleLockScreen);
+    electron.powerMonitor.off('unlock-screen', this.handleUnlockScreen);
+
+    // stop monitoring system idle time
+    clearInterval(this.systemIdleTimeIntervalID);
+  }
+
+  // - updates internal idle time properties and conditionally triggers various tasks based on those values
+  updateIdleTime(idleTime) {
+    // store last and this idleTime for comparison
+    this.lastIdleTime = this.idleTime;
+    this.idleTime = idleTime;
+
+    // provide frequent activity updates
+    this.emit('update', {userIsActive: this.isActive, idleTime: this.idleTime});
+
+    if (this.isActive && this.idleTime > this.config.userActivityTimeout) {
```

Comment:
I think these checks would be easier if we dispense with the `this.isActive` checks -- we already check the value in `updateUserActivityStatus` to avoid emitting spurious events anyway.

The last else condition, then, could be moved into `updateUserActivityStatus` instead, having it emit `sendStatusUpdate` /anyway/ even if the value hasn't changed. Or even `sendStatusUpdate` itself could debounce itself to avoid anyone calling it more often than needed?

Reply:
Reorganized, thanks.